jbmc/src/java_bytecode/README.md

```diff
@@ -736,3 +736,151 @@ See also \ref string-solver-interface.
+assigning an entire array to the `java::array` symbol instead of being built
+using index assignments.
+See \ref java-trace-array-assignments "Array assignments".
+
+\subsection java-trace-variable-assignment Variable assignments
+
+Variable assignments have LHS \ref symbol_exprt with an identifier that reflects
+the variable name and scope and a RHS that is one of the following
+(noting that \ref java-trace-array-assignment Strings are an exception,
+see \ref java-trace-string-assignments "String assignments"):
+- \ref constant_exprt (primitives and addresses)
+- \ref struct_exprt (reference types, including Java arrays)
+- \ref array_exprt (data within Java arrays)
+
+A primitive variable assignment has the form:
+- \ref symbol_exprt = \ref const_exprt with a value representing the primitive
+
+An assignment of a non-primitive (struct) variable can have the following forms:
+- \ref symbol_exprt = \ref struct_exprt
+- \ref symbol_exprt = \ref const_exprt with an \ref address_of_exprt pointing to a symbol that has a \ref struct_exprt value
+
+\subsection java-trace-member-assignments Member assignments
+
+\ref struct_exprt that have members are constructed first by an assignment of
+the form
+- \ref symbol_exprt = \ref struct_exprt
+
+where the members of the \ref struct_exprt are zero-intialized (null for
+\ref struct_exprt and 0 for \ref constant_exprt)
+then members can be assigned. For primitive members, this is a simple assignment
+to the member:
+- \ref member_exprt of the \ref symbol_exprt containing the member = \ref constant_exprt with a value representing the primitive
+
+For members of reference type, the object to be assigned must first be created,
+then assigned by address, e.g.
+- \ref member_exprt of the \ref symbol_exprt containing the member = \ref const_exprt with an \ref address_of_exprt pointing to a symbol that has a \ref struct_exprt value.
+
+This concept becomes clearer with an example.
+
+\subsection java-trace-member-assignment-example Worked example for member assignments
+
+Consider the trace for the function call to `Example.assignReferenceTypes()`,
+which assigns a value to the static field `globalReferenceTypeReferenceField`.
+
+```java
+public class Example {
+  static ReferenceTypeReferenceField globalReferenceTypeReferenceField;
+  public static void assignReferenceTypes() {
+    globalReferenceTypeReferenceField =
+        new ReferenceTypeReferenceField(new ReferenceTypePrimitiveField(5));
+}
+
+class ReferenceTypeReferenceField {
+  ReferenceTypePrimitiveField referenceField;
+  public ReferenceTypeReferenceField(ReferenceTypePrimitiveField r) {
+    referenceField = r;
+  }
+}
+
+class ReferenceTypePrimitiveField {
+  int primitiveField;
+  public ReferenceTypePrimitiveField(int i) {
+    primitiveField = i;
+  }
+}
+```
+
+There are 54 non-trivial steps in the trace to make this assignment.
+This example will give detail of the assignment steps only, filling in other
+steps with narrative.
```

Comment:
❓ What do you mean by "narrative" here?

Reply:
I mean `description` instead of writing out the assignment in pseudo-code. (I was free-writing to try to get this done today, so I chose a stupid word and didn't change it :) )